cpp/src/parquet/types.h

```diff
@@ -89,11 +90,831 @@ struct LogicalType {
+  inline bool is_list(void) const { return type_ == LogicalAnnotation::Type::List; }
+  inline bool is_enum(void) const { return type_ == LogicalAnnotation::Type::Enum; }
+  inline bool is_decimal(void) const { return type_ == LogicalAnnotation::Type::Decimal; }
+  inline bool is_date(void) const { return type_ == LogicalAnnotation::Type::Date; }
+  inline bool is_time(void) const { return type_ == LogicalAnnotation::Type::Time; }
+  inline bool is_timestamp(void) const {
+    return type_ == LogicalAnnotation::Type::Timestamp;
+  }
+  inline bool is_interval(void) const {
+    return type_ == LogicalAnnotation::Type::Interval;
+  }
+  inline bool is_int(void) const { return type_ == LogicalAnnotation::Type::Int; }
+  inline bool is_null(void) const { return type_ == LogicalAnnotation::Type::Null; }
+  inline bool is_JSON(void) const { return type_ == LogicalAnnotation::Type::JSON; }
+  inline bool is_BSON(void) const { return type_ == LogicalAnnotation::Type::BSON; }
+  inline bool is_UUID(void) const { return type_ == LogicalAnnotation::Type::UUID; }
+  inline bool is_none(void) const { return type_ == LogicalAnnotation::Type::None; }
+  //
+  inline bool is_valid(void) const { return type_ != LogicalAnnotation::Type::Unknown; }
+  inline bool is_nested(void) const {
+    return (type_ == LogicalAnnotation::Type::List) ||
+           (type_ == LogicalAnnotation::Type::Map);
+  }
+  inline bool is_serialized(void) const {
+    return !((type_ == LogicalAnnotation::Type::None) ||
+             (type_ == LogicalAnnotation::Type::Unknown));
+  }
+  inline bool is_nonnested(void) const { return !is_nested(); }
+  inline bool is_invalid(void) const { return !is_valid(); }
+
+  // Return the appropriate sort order of this annotation.
+  inline SortOrder::type sort_order(void) const { return order_; }
+
+  virtual ~LogicalAnnotation() {}
+
+ protected:
+  LogicalAnnotation(LogicalAnnotation::Type t, SortOrder::type o) : type_(t), order_(o) {}
+
+ private:
+  LogicalAnnotation() {}
+
+  LogicalAnnotation::Type type_ = LogicalAnnotation::Type::Unknown;
+  SortOrder::type order_ = SortOrder::UNKNOWN;
+};
+
+class PARQUET_EXPORT StringAnnotation : public LogicalAnnotation {
+ public:
+  inline bool is_applicable(const parquet::Type::type primitive_type,
+                            const int32_t primitive_length = -1) const override {
+    return primitive_type == parquet::Type::BYTE_ARRAY;
+  }
+
+  inline bool is_compatible(
+      const LogicalType::type converted_type,
+      const schema::DecimalMetadata converted_decimal_metadata) const override {
+    return (converted_type == LogicalType::UTF8) && !converted_decimal_metadata.isset;
+  }
+
+  inline bool has_converted_type(void) const override { return true; }
+
+  inline LogicalType::type ToConvertedType(
+      schema::DecimalMetadata* converted_decimal_metadata) const override {
+    reset_decimal_metadata(converted_decimal_metadata);
+    return LogicalType::UTF8;
+  }
```

Comment:
I tinkered around early on with a few ideas like this (generally trying to capture commonalities, reduce very similar code) but it appeared at the time that the result would be somewhat over-engineered.  However the code has evolved and settled some since then, so I'm going to take another pass at that.